### PR TITLE
vim-patch: 8.0.x, 8.1.x menu features

### LINKED ISSUE #14921: mousemodel=popup not working
- `nvim --version`: NVIM v0.5.0-dev+1449-gbdf3df402-dirty
- Operating system/version: Ubuntu 21.04
- Terminal name/version: Konsole 20.12.3
- `$TERM`: xterm-256color

[X] `vim -u DEFAULTS` (version: 8.2.2434) behaves differently

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
```
```vim
:set mouse=a
:set mousemodel=popup
:menu PopUp.Save :w<CR>
```

* Right click on the screen

### Actual behaviour

No response

### Expected behaviour

A popup menu shown with a `Save` item

### Additional information

`:help 'mousemodel'` suggests that a setting of `popup` would result in a popup menu when right-clicking.

```
	   popup	Right mouse button pops up a menu.  The shifted left
			mouse button extends a selection.  This works like
			with Microsoft Windows.
```

### PR DESCRIPTION
**Pull request by zeertzjq**
Fix #14921

#### vim-patch:8.0.1558: no right-click menu in a terminal

Problem:    No right-click menu in a terminal.
Solution:   Implement the right click menu for the terminal.
https://github.com/vim/vim/commit/aef8c3da2ba59285b7cfde559ae21cdce6ba6919

#### vim-patch:8.0.1570: can't use :popup for a menu in the terminal

Problem:    Can't use :popup for a menu in the terminal. (Wei Zhang)
Solution:   Make :popup work in the terminal.  Also fix that entries were
            included that don't work in the current state.
https://github.com/vim/vim/commit/29a2c08d792e4458a0af8371f5341394829fce29

#### vim-patch:8.0.1574: show cursor in wrong place when using popup menu

Problem:    Show cursor in wrong place when using popup menu. (Wei Zhang)
Solution:   Force updating the cursor position.  Fix skipping over unused
            entries.
https://github.com/vim/vim/commit/987723e084660290270b3c3d943eb13bd828d5da

#### vim-patch:8.0.1588: popup menu hangs after typing CTRL-C

Problem:    Popup menu hangs after typing CTRL-C.
Solution:   Make CTRL-C exit the loop. (Ozaki Kiichi, closes vim/vim#2697)
https://github.com/vim/vim/commit/52f18a112a073c39187cd93f26115d1e2bfd29ce

#### vim-patch:8.1.0487: no menus specifically for the terminal window

Problem:    No menus specifically for the terminal window.
Solution:   Add :tlmenu. (Yee Cheng Chin, closes vim/vim#3439) Add a menu test.
https://github.com/vim/vim/commit/4c5d815256099b50eca2ec5bf8f9aaa67a890211

ADDR_OHTER comes from patch 8.1.1241, which has already been ported.

#### vim-patch:8.1.0695: internal error when using :popup

Problem:    Internal error when using :popup.
Solution:   When a menu only exists in Terminal mode give an error. (Naruhiko
            Nishino, closes vim/vim#3765)
https://github.com/vim/vim/commit/f42b45d719e03218735b3c2845a74dca9c0efd60

#### vim-patch:8.1.1274: after :unmenu can still execute the menu with :emenu

Problem:    After :unmenu can still execute the menu with :emenu.
Solution:   Do not execute a menu that was disabled for the specified mode.
https://github.com/vim/vim/commit/ce79353ace9e21238f13655089363cd23cbb6b32

### CODE REVIEW

**Comment by zeertzjq** [comment]
Adding tests will take some time. I don't plan to make right-click popup menu to support `ext_popupmenu`  or event processing in this PR.

File: runtime/doc/index.txt [hunk 1]
```diff
@@ -1622,6 +1622,9 @@ tag		command		action ~
 |:tjump|	:tj[ump]	like ":tselect", but jump directly when there
 				is only one match
 |:tlast|	:tl[ast]	jump to last matching tag
+|:tlmenu|	:tlm[enu]	add menu for Terminal-Job mode
+|:tlnoremenu|	:tln[oremenu]	like ":noremenu" but for Terminal-Job mode
+|:tlunmenu|	:tlu[nmenu]	remove menu for Terminal-Job mode
```
Comment (justinmk) [inline runtime/doc/index.txt:1627]:
```suggestion
|:tlmenu|	:tlm[enu]	add menu for |Terminal-mode|
|:tlnoremenu|	:tln[oremenu]	like ":noremenu" but for |Terminal-mode|
|:tlunmenu|	:tlu[nmenu]	remove menu for |Terminal-mode|
```

File: test/functional/ui/popupmenu_spec.lua [hunk 2]
```diff
@@ -2359,6 +2360,74 @@ describe('builtin popupmenu', function()
       {2:-- INSERT --}                    |
     ]])
   end)
+
+  it('supports mousemodel=popup', function()
```
Comment (justinmk) [inline test/functional/ui/popupmenu_spec.lua:2364]: tried it out locally, it works! :yayyyyyyy:

File: src/nvim/menu.c [hunk 3]
```diff
@@ -1480,12 +1615,60 @@ void ex_emenu(exarg_T *eap)
...
     return;
   }
 
   // Found the menu, so execute.
-  execute_menu(eap, menu);
+  execute_menu(eap, menu, mode_idx);
+}
+
+/// Given a menu descriptor, e.g. "File.New", find it in the menu hierarchy.
+vimmenu_T *gui_find_menu(const char *path_name)
```
Comment (justinmk) [inline src/nvim/menu.c:1629]:
```suggestion
vimmenu_T *menu_find(const char *path_name)
```

**Review by justinmk — APPROVED** [review]

**Pull request merged**